DiceDB/dice
# Dice Nitro Server (Multi threaded server) Implementation

- Added complete wiring & implementation for nitro server.
- Wired test setup to use Nitro server, behind a feature flag DICE_TEST_NITRO_SERVER_CORES
- Fix integration tests with nitro server
- Added make commands for nitro server testing

How to run Nitro Server?
Just pass --cores parameter to existing main.go. 

```
go run main.go --cores 50
```

How to test Nitro Server?

```
export DICE_TEST_NITRO_SERVER_CORES=3 # or whatever
make test
```
You can also use the make command `make test-nitro`

Current Open Issues
- (Blocker) QWATCH is not functioning as expected.
- Existing Unit tests are unaffected
- Existing Integration are unaffected
- (Non-Blocker) Many integration tests are still failing on Nitro mode

Proposal:
Propose to review the overall structure and address any critical concerns. We can merge the current implementation with a few integration tests failing for Nitro mode. We can create new tickets for failing tests in nitro mode and divide and conquer them.

Dice CLI - 
<img width="1680" alt="Screenshot 2024-08-22 at 6 42 12 PM" src="https://github.com/user-attachments/assets/d31429e3-8553-43ba-b9b7-beccfceabcf8">


Sample Integration Test with new Make command- 

```
 ~/work/code/dice   blue/multithreaded-server  TEST_FUNC=TestGetEx make test-one-nitro
export DICE_TEST_NITRO_SERVER_CORES=3 && go test -v -count=1 --run TestGetEx ./tests/... && unset DICE_TEST_NITRO_SERVER_CORES
starting the test server on port 8739
2024/08/22 18:44:42 INFO Starting Nitro Test Server with Cores= 3="missing value"
2024/08/22 18:44:42 INFO Starting Dice Nitro server on 0.0.0.0=8739
2024/08/22 18:44:42 INFO Initializing Shard pool with Cores = 5
2024/08/22 18:44:42 INFO Nitro server is ready to accept connections
=== RUN   TestGetEx
=== RUN   TestGetEx/GetEx_Simple_Value
=== RUN   TestGetEx/GetEx_Non-Existent_Key
=== RUN   TestGetEx/GetEx_with_EX_option
=== RUN   TestGetEx/GetEx_with_PX_option
=== RUN   TestGetEx/GetEx_with_EX_option_and_invalid_value
=== RUN   TestGetEx/GetEx_with_PX_option_and_invalid_value
=== RUN   TestGetEx/GetEx_with_EXAT_option
=== RUN   TestGetEx/GetEx_with_PXAT_option
=== RUN   TestGetEx/GetEx_with_EXAT_option_and_invalid_value
=== RUN   TestGetEx/GetEx_with_PXAT_option_and_invalid_value
=== RUN   TestGetEx/GetEx_with_Persist_option
=== RUN   TestGetEx/GetEx_with_multiple_expiry_options
=== RUN   TestGetEx/GetEx_with_persist_and_ex_options
=== RUN   TestGetEx/GetEx_with_persist_and_px_options
--- PASS: TestGetEx (14.02s)
    --- PASS: TestGetEx/GetEx_Simple_Value (0.00s)
    --- PASS: TestGetEx/GetEx_Non-Existent_Key (0.00s)
    --- PASS: TestGetEx/GetEx_with_EX_option (2.00s)
    --- PASS: TestGetEx/GetEx_with_PX_option (2.00s)
    --- PASS: TestGetEx/GetEx_with_EX_option_and_invalid_value (0.00s)
    --- PASS: TestGetEx/GetEx_with_PX_option_and_invalid_value (0.00s)
    --- PASS: TestGetEx/GetEx_with_EXAT_option (5.00s)
    --- PASS: TestGetEx/GetEx_with_PXAT_option (5.00s)
    --- PASS: TestGetEx/GetEx_with_EXAT_option_and_invalid_value (0.00s)
    --- PASS: TestGetEx/GetEx_with_PXAT_option_and_invalid_value (0.00s)
    --- PASS: TestGetEx/GetEx_with_Persist_option (0.00s)
    --- PASS: TestGetEx/GetEx_with_multiple_expiry_options (0.00s)
    --- PASS: TestGetEx/GetEx_with_persist_and_ex_options (0.00s)
    --- PASS: TestGetEx/GetEx_with_persist_and_px_options (0.00s)
PASS
2024/08/22 18:44:57 INFO Context closed
2024/08/22 18:44:57 INFO Context closed
ok      github.com/dicedb/dice/tests    15.797s
```
